Step3 자동차 경주 구현

gongwho opened the pull request:
- 기능 구현
- 리뷰 반영

silvio-oivlis committed:
study: 문자열 계산기 구현 및 테스트코드 작성

silvio-oivlis committed:
study: 자동차 경주 시뮬레이터 생성

silvio-oivlis committed:
study: 자동차 경주 시뮬레이터 생성

silvio-oivlis committed:
study: 리뷰 반영 (operator 분리, enum 적용)

silvio-oivlis committed:
study: 리뷰 반영 (test 에 Matcher 적용)

silvio-oivlis committed:
study: 자동차 경주 구현 및 리뷰 반영

silvio-oivlis committed:
resolve merge conflicts

silvio-oivlis committed:
add residue codes

wisemuji commented on src/main/kotlin/StringCalculator.kt:
```diff
@@ -9,29 +9,34 @@ object StringCalculator {
+    }
 
-    fun divide(currentValue: Int, operand: Int) = currentValue / operand
+    object Operator {
+        fun plus(currentValue: Int, operand: Int) = ArithmeticOperation.Plus.run(currentValue, operand)
+        fun minus(currentValue: Int, operand: Int) = ArithmeticOperation.Minus.run(currentValue, operand)
+        fun multiply(currentValue: Int, operand: Int) = ArithmeticOperation.Multiply.run(currentValue, operand)
+        fun divide(currentValue: Int, operand: Int) = ArithmeticOperation.Divide.run(currentValue, operand)
+        fun performOperation(currentValue: Int, operation: ArithmeticOperation, operand: Int): Int =
+            operation.run(currentValue, operand)
+    }
```
테스트에서만 활용되는 코드가 실제 main 디렉토리에 포함되는 것이 적절할까요?
자동차 경주 프로그램을 만드는 데에 집중하실 수 있도록 문자열 계산기에 대한 피드백은 이만 줄이겠습니다 🙂

gongwho replied:
이렇게 작성하게 된 것은 몇가지 의도와 이유가 있었는데요!
- 수정 중 step3 진행하느라 변경 중단
- 리팩토링 과정에서 기존 코드와의 차이를 줄이고 싶은 의도(이 경우 plus(Int, Int) 와 같은 public 함수 signature)
- StringCalculator 가 아닌 ***Calculator 를 작성하게 되는 경우를 고려해 Calculator > Operator > ArithmeticOperation 으로 논리적 의존성 설계
- ArithmeticOperation 이 아닌 기능(power, sqrt, factorial 등) 고려하던 중 Operator 의 기능적 일관성 소실

고민은 많았지만, 결과적으로 저도 썩 만족스럽지 않은 코드가 된 것 같습니다 ㅎㅎ
짚어주지 않으셨으면 고민을 이어가지 못했을 수 있을것 같아서 꼼꼼히 봐주신 점 감사드립니다 :)

리뷰를 반영하고 수정할 여력이 없지 않아서, 괜찮으시면 이후 수정에 대해서도 리뷰해주실 수 있을까요?

wisemuji commented on src/test/kotlin/study/CarRaceSimulatorTest.kt:
```diff
@@ -0,0 +1,10 @@
+import CarRaceSimulator
+import io.kotest.core.spec.style.StringSpec
+
+class CarRaceSimulatorTest : StringSpec({
+    "carRaceSimulatorTest" {
+        CarRaceSimulator.simulate(CarRaceSimulator.CarListProvideMethod.Manual(3, 50))
+    }
+})
```
유효하지 않은 테스트인것으로 보입니다.
어떤 테스트가 의미있게 기능을 검증할 수 있는 테스트일까요?

gongwho replied:
혹시 이렇게 자동화 목적이 아닌 테스트 코드를 작성하게 되는 경우가 있을까요?

현업에서 main 함수를 실행 하는 것이 무거운 작업인 경우가 잦을것 같아 저비용으로 결과만 눈으로 확인하려고 작성한 코드였습니다,,

더 적절한 방법이 있다면 저도 목적에 맞게 변경하고 싶은데, 이해가 짧아서 찾지 못한것 같습니다,,

wisemuji replied:
@gongwho 홍구님이 말씀해주신 목적에 걸맞는 테스트 코드도 충분히 작성할 수 있습니다 🙂  
단, 이번 미션에서는 전체 프로그램에 대한 테스트보다 각각의 기능을 단위로 쪼개어 검증하는 테스트 작성부터 연습해보면 어떨까요?
미션을 조금씩 진행하시고 수업 시간에 작성되는 코드를 보며 테스트에 대한 재미를 얻어가실 수 있을 것이라고 믿습니다.

wisemuji commented on src/main/kotlin/CarRaceSimulator.kt:
```diff
@@ -0,0 +1,51 @@
+
+    data class Car(val traveled: Int)
+
+    class CarList(private val cars: MutableList<Car>) {
+
+        fun driveCars() {
+            for (n in 0 until cars.size) {
+                cars[n] = cars[n].drive()
+            }
+        }
+
+        fun printCars() {
+            cars.forEach { car ->
+                car.print()
+            }
+            println()
+        }
+
+        private fun Car.drive() = if ((0..9).random() >= 4) {
+            copy(traveled = traveled + 1)
+        } else this
+
+        private fun Car.print() = println("-".repeat(traveled))
+    }
```
미션 요구사항에 맞춰 분리해보면 어떨까요?
> 핵심 로직을 구현하는 코드와 UI를 담당하는 로직을 구분한다.

InputView, ResultView와 같은 별도의 클래스로 나눠보면 좋겠습니다!

wisemuji commented on src/main/kotlin/CarRaceSimulator.kt:
```diff
@@ -0,0 +1,51 @@
+        }
+    }
+
+    sealed interface CarListProvideMethod {
+        fun provide(): Pair<Int, Int>
+        class Manual(private val carCount: Int, private val iterationCount: Int) : CarListProvideMethod {
+            override fun provide(): Pair<Int, Int> = Pair(carCount, iterationCount)
+        }
+
+        object UserInput : CarListProvideMethod {
+            override fun provide(): Pair<Int, Int> {
+                val carCount = println("자동차 대수는 몇 대인가요?").run { readln().trim().toInt() }
+                val iterationCount = println("시도할 횟수는 몇 회인가요?").run { readln().trim().toInt() }
+                return Pair(carCount, iterationCount)
+            }
+        }
+    }
+
```
CarRaceSimulator 하위에 CarListProvideMethod와 Car, CarList 등의 [nested class](https://kotlinlang.org/docs/nested-classes.html#inner-classes)로 만들어주셨는데요, 
꼭 nested class로 만들어야 할 이유가 있을까요?

gongwho replied:
반대로 ... 굳이 지금 패키지를 나누거나 별개의 파일을 만들 필요가 있을까 싶어서 내버려뒀었습니다 ㅎㅎ
좀 더 생각해보니 미리 작업해 두는 것이 더 적절할 것 같네요!

wisemuji commented on src/main/kotlin/CarRaceSimulator.kt:
```diff
@@ -0,0 +1,51 @@
+object CarRaceSimulator {
+    fun simulate(provideMethod: CarListProvideMethod) {
+        val (carCount, iterationCount) = provideMethod.provide()
+
+        val carList = CarList(MutableList(carCount) { Car(0) })
+
+        for (i in 0 until iterationCount) {
+            carList.driveCars()
+            carList.printCars()
+        }
+    }
```
다른 개발자들이 자동차 경주 프로그램을 실행할 수 있도록 main 함수를 추가해보면 어떨까요?

wisemuji commented on src/main/kotlin/CarRaceSimulator.kt:
```diff
@@ -0,0 +1,51 @@
+
+    sealed interface CarListProvideMethod {
+        fun provide(): Pair<Int, Int>
+        class Manual(private val carCount: Int, private val iterationCount: Int) : CarListProvideMethod {
+            override fun provide(): Pair<Int, Int> = Pair(carCount, iterationCount)
+        }
```
테스트 코드를 위한 작업인걸까요? 
테스트에서만 활용되는 코드가 실제 main 디렉토리에 포함되는 것이 적절할까요?

gongwho replied:
결과적으로 테스트 코드에서만 활용하고 있지만, input 을 받지 않는 provision 방식은 반드시 필요할 것 같아 작성했습니다!

wisemuji commented on src/main/kotlin/CarRaceSimulator.kt:
```diff
@@ -0,0 +1,51 @@
+            println()
+        }
+
+        private fun Car.drive() = if ((0..9).random() >= 4) {
+            copy(traveled = traveled + 1)
+        } else this
```
이러한 로직은 어떻게 테스트할 수 있을까요?

wisemuji reviewed (changes requested):
안녕하세요 홍구님!

Step3 구현을 잘 해주셨습니다 👍 
고민해보시면 좋을만한 코멘트를 달았으니 확인해주시면 감사하겠습니다.
이번 단계에서는 다음 요구사항을 만족하는데에 집중해보면 어떨까 싶습니다.
> 핵심 로직을 구현하는 코드와 UI를 담당하는 로직을 구분한다.

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏 💪

wisemuji reviewed (changes requested):
안녕하세요 홍구님!

Step3 구현을 잘 해주셨습니다 👍 
고민해보시면 좋을만한 코멘트를 달았으니 확인해주시면 감사하겠습니다.
이번 단계에서는 다음 요구사항을 만족하는데에 집중해보면 어떨까 싶습니다.
> 핵심 로직을 구현하는 코드와 UI를 담당하는 로직을 구분한다.

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏 💪

silvio-oivlis committed:
study: 자동차 경주 피드백 반영

- 전체적인 패키지 구조 정리
- StringCalculator 리팩토링
- CarRace 미션 요구사항 반영
- CarRace 테스트 코드 작성
- CarRaceSimulator main 함수 작성

wisemuji commented on src/main/kotlin/carRace/ResultView.kt:
```diff
@@ -0,0 +1,11 @@
+        println()
+    }
+    private fun Car.print() = println("-".repeat(traveled))
+}
```
다음 설정 옵션을 활성화하여 EOF 경고를 없애주세요 🙂

Preferences > Editor > General > Ensure every saved file ends with a line break

Ktlint 체크도 실패하고 있습니다. 다음 명령어로 깨지는 린트를 확인해주세요.
```
./gradlew check
```

gongwho replied:
IDE 에 문제가 있는지 제가 실행할땐 check, ktlintCheck 모두 통과하고 있었네요 ㅠ 다음 커밋에 반영하겠습니다~

wisemuji commented on src/main/kotlin/calculator/Operation.kt:
```diff
@@ -0,0 +1,7 @@
+package calculator
+
+sealed interface Operation
+
+sealed interface BinaryOperation : Operation {
+    fun run(operand1: Int, operand2: Int): Int
+}
```
sealed interface와 enum을 혼용해서 사용하기보다 이번 기회에 각각의 차이점을 이해하고 어떤 상황에서 사용하면 좋을지 판단해보면 어떨까요?

- https://blog.kotlin-academy.com/enum-vs-sealed-class-which-one-to-choose-dc92ce7a4df5

wisemuji commented on src/main/kotlin/carRace/RaceRegulationProvider.kt:
```diff
@@ -0,0 +1,18 @@
+
+sealed interface RaceRegulationProvider {
+
+    fun provide(): Pair<Int, Int>
```
Pair 타입은 사용할 땐 굉장히 편하지만 의미를 유추하기 어렵다는 단점이 있습니다. 다른 개발자가 이 코드를 봤을 때 각각의 Int 타입이 무엇을 의미하는지 예측할 수 있을까요? 새로운 클래스를 만드는 것을 너무 어렵게 생각하지 않아도 됩니다.

wisemuji commented on src/main/kotlin/calculator/StringCalculator.kt:
```diff
@@ -0,0 +1,33 @@
+    private fun getOperation(operationString: String): ArithmeticOperation {
+        return ArithmeticOperation.values().firstOrNull {
+            it.operationString == operationString
+        } ?: throw IllegalArgumentException("invalid operation: `$operationString`")
```
ArithmeticOperation을 외부에서 꺼내 사용하기보다 메시지를 던져보면 어떨까요?

wisemuji commented on src/main/kotlin/carRace/Car.kt:
공식 문서에 따르면 동일한 코틀린 파일에 여러 클래스를 두는 것은 이들이 서로 밀접하게 연관되어 있고 파일 크기가 합리적인 상태를 유지할 때에만 권장됩니다. 한 파일에 두는 특별한 이유가 없다면 별도의 파일로 분리하면 어떨까요?

- https://kotlinlang.org/docs/coding-conventions.html#source-file-organization

wisemuji commented on src/main/kotlin/carRace/Car.kt:
```diff
@@ -0,0 +1,21 @@
+data class Car(private var _traveled: Int, val carMovingStrategy: CarMovingStrategy = RandomMovingStrategy()) {
+    private fun shouldMove(): Boolean = carMovingStrategy.shouldMove()
+
+    val traveled: Int get() = _traveled
```
코틀린 코딩 컨벤션에서는 클래스 내부 요소들의 정렬을 다음과 같이 정의하고 있습니다.
1. 프로퍼티 선언 및 초기화 블록
2. 보조 생성자
3. 함수
4. companion object
- https://kotlinlang.org/docs/coding-conventions.html#class-layout

wisemuji replied:
[backing property](https://kotlinlang.org/docs/properties.html#backing-properties)를 활용하여 가변 값을 관리하도록 구현하셨군요 👍 
다음과 같이 작성할수도 있으니 참고만 해주세요.

```kotlin
var traveled: Int
    private set
```

> If you need to annotate an accessor or change its visibility, but you don't want to change the default implementation, you can define the accessor without defining its body:

wisemuji commented on src/main/kotlin/carRace/CarList.kt:
```diff
@@ -0,0 +1,9 @@
+package carRace
+
+class CarList(val cars: MutableList<Car>) {
```
MutableList일 필요가 있을까요?
많은 코틀린의 타입은 불변이 기본값이고 List도 그중 하나입니다. 이런 불변 타입을 사용하는 것의 장점을 온전히 누리기 위해 전 코틀린에서는 가능하면 var이나 Mutable 타입 사용을 지양하려고 하고 있습니다. :slightly_smiling_face:

wisemuji commented on src/main/kotlin/carRace/CarRaceSimulator.kt:
```diff
@@ -0,0 +1,22 @@
+
+        val carList = CarList(MutableList(carCount) { Car(0, carMovingStrategy) })
+
+        for (i in 0 until iterationCount) {
+            carList.driveCars()
+        }
```
[repeat](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/repeat.html)을 활용해보면 어떨까요?

wisemuji commented on src/test/kotlin/study/CarRaceSimulatorTest.kt:
```diff
@@ -0,0 +1,54 @@
+import io.kotest.matchers.shouldBe
+
+class CarRaceSimulatorTest : StringSpec({
+    "carDriveTest" {
```
테스트명을 통해 어떤 것을 테스트하고자 하는 것이 제대로 드러나고 있지 않습니다. 테스트명을 작성하실 때에는
1. 검증하고자 하는 로직을 명확하게 정하고 (검증이 필요한 로직이 두 개 이상이면 필요에 따라 또 다른 함수로 분할한다)
2. 그것을 함수명으로 드러내보면 어떨까요?

wisemuji replied:
다음 미션을 진행하실때는 기능을 구현하기 전에 README.md 파일에 구현할 기능 목록을 정리해 추가해보면 어떨까요?

테스트명을 가장 쉽게 만들 수 있는 방법 중 하나는 README.md에 추가한 기능 요구사항을 그대로 테스트로 옮겨오는 것입니다. 커밋 메시지를 작성할 때에도 동일하고요!

wisemuji commented on src/test/kotlin/study/CarRaceSimulatorTest.kt:
```diff
@@ -0,0 +1,54 @@
+import io.kotest.core.spec.style.StringSpec
+import io.kotest.matchers.shouldBe
+
+class CarRaceSimulatorTest : StringSpec({
```
Simulator를 테스트하기 위해 전략 패턴 등을 활용하여 테스트하기 어려운 영역을 분리해주셨네요 👍 

5단계를 진행하시면 더 와닿으시겠지만 테스트 코드를 만들 때에는 테스트하기 쉬운 영역에 해당하는, 핵심 비즈니스 로직을 갖는 로직 위주로 테스트를 작성하는 것부터 시작합니다.

Car 클래스에도 단위 테스트를 추가해보면 어떨까요?

wisemuji reviewed (approved):
안녕하세요 홍구님! 
리뷰 반영을 잘 해주셨습니다 💯 

조금 고민해보시면 좋을 내용을 코멘트로 달았으니 다음 단계 미션 진행해주시면서 같이 반영해주시면 됩니다!

다음 미션도 화이팅입니다 🙏
